perf: memo SidebarLinks

hannojg opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

When clicking the `+` FAB on the start screen I felt that it took too long until the popup opens. I want the popup to open instantly. After some performance profiling, I've seen that when opening the popup the `SidebarLinks` unnecessarily rerender. I assume that fix also maybe helps in other places where it would rerender although it doesn't need to.

See a video before and after the change:

| **Before** | **After** |
|--|--|
| <video src="https://user-images.githubusercontent.com/16821682/196205400-e5a4746a-f32c-477b-b166-d5a1b44dad8f.mp4" /> | <video src="https://user-images.githubusercontent.com/16821682/196205539-cd3c876e-09f3-4753-9829-391bc95585dd.mp4" /> |

> Note: with this change I was able to witness that the FAB actually has an animation when pressing it 👀 😄 (before the frames were skipped)

(A performance comparison from debug mode):
<img width="1278" alt="perf-comparison" src="https://user-images.githubusercontent.com/16821682/196207171-659941f7-a937-49e5-8401-be3b701e838b.png">


### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/12119
PROPOSAL: GH_LINK_ISSUE(COMMENT)


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Click the + FAB on the HomeScreen
2. Press the emoji picker in a chat
3. Press the "attachment" icon in the chat composer
4. Verify that its possible to open and close all these modals flawlessly without any jank (You should e.g. see the animation on the FAB button)

- [x] Verify that no errors appear in the JS console

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The reviewer will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
5. Upload an image via copy paste
6. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->

https://user-images.githubusercontent.com/16821682/196210965-0b3ced95-881d-4ff6-81ad-f955ca4d0ff5.mov


#### Mobile Web - Chrome
<!-- Insert screenshots of your changes on the web platform (from chrome mobile browser)-->

https://user-images.githubusercontent.com/16821682/196211036-a15b9c20-91e9-4f40-b71f-55b4b18324ed.mov


#### Mobile Web - Safari
<!-- Insert screenshots of your changes on the web platform (from safari mobile browser)-->

Couldn't test due to this bug appearing: https://github.com/Expensify/App/issues/11677

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

https://user-images.githubusercontent.com/16821682/196211116-a8a28f63-d64d-42d0-a406-e91638f65836.mov


#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

https://user-images.githubusercontent.com/16821682/196211180-797b21d9-9dda-4fa9-b90f-8a162b62f58f.mov


#### Android
<!-- Insert screenshots of your changes on the Android platform-->

https://user-images.githubusercontent.com/16821682/196211256-eb67da25-e323-41fe-a1af-26fe312e35f2.mp4

hannojg committed:
move platform code to specific functions

hannojg committed:
clean code

hannojg committed:
perf: memo SidebarLinks

mountiny reviewed (commented):
@hannojg Thank you very much for this change, looks like great improvement.

Could you please write down a set of Tests and QA steps for the Applause QA team which they could use to make sure this does not have any unintentional sideeffects?

mountiny commented on src/pages/home/sidebar/SidebarLinks.js:
```diff
@@ -164,36 +164,38 @@ class SidebarLinks extends React.Component {
-    withCurrentUserPersonalDetails,
-    withWindowDimensions,
-    withOnyx({
+export default React.memo(
```
Additionally, can you please leave a comment why to use this, basically another question woul dbe why to use it here and not at other places

hannojg replied:
Yeah so, `SidebarLinks` is rendered in `BaseSidebarScreen`, which also renders the `PopoverMenu`. So when the `isCreateMenuActive` state changes it rerenders the `PopoverMenu` which is desired, but also the `SidebarLinks`, which doesn't need to rerender:

https://github.com/Expensify/App/blob/6db559f9b42dbfa5e1e09e60d692272d85039e9c/src/pages/home/sidebar/SidebarScreen/BaseSidebarScreen.js#L112-L129

It's important to wrap the Function returned by `compose`. This way none of the HoCs has to run. Especially `withOnyx` is expensive to run. That's why we don't just turn e.g. the class to a `PureComponent`, or wrap the Class with `React.memo`, but really wrap the whole thing

hannojg replied:
Another solution here could be to move the `PopoverMenu` with the `isVisible` state into its own component, so only that component rerenders and not the whole screen.

mountiny replied:
Oh yeah, thank you for this, I mean to add a comment to the code for any engineer wondering.

Also I feel like reading about the React memo, what is curious to me is why not to use it in more places? Or did you not find other place which would have this rendering issue?

tgolen replied:
@mountiny `React.memo` should be used pretty sparingly because it can have some pretty unintentional side-effects. It can be a bit lazy to just use it everywhere, thinking it'll be a good fix. I believe, primarily, it only does a shallow comparison of props, which isn't always fully understood.

> Another solution here could be to move the PopoverMenu with the isVisible state into its own component, so only that component rerenders and not the whole screen.

I'd actually much rather do this than use `React.memo` in this PR. I think a lot of the sidebar re-rendering problems will go away when I am able to apply the Onyx property selectors to it (see more details of what this is about in [this GH](https://github.com/Expensify/Expensify/issues/222430)).

mountiny replied:
Great, yeah, let's definitely make new PRs for each refactor.

hannojg commented:
@mountiny Hey, I was thinking about steps, but having a bit of a hard time coming up with some (I think it shouldn't cause any strange side effects, but totally agree its good to test that as much as possible!).  I am not too familiar yet with the app functionality and looking more at the app through the lens of performance profiling tools. Maybe there is someone who is more familiar with the `SidebarLinks` component and could come up with some suggestions?

mountiny commented:
> Maybe there is someone who is more familiar with the SidebarLinks component and could come up with some suggestions?

Yeah I think general regression tests would be good, but in term of SidebarLinks specifically, @tgolen is the best at the moment I think.

tgolen reviewed (dismissed):
Is there a GH issue for this PR? It looks like this section is not fully filled out:

![image](https://user-images.githubusercontent.com/1228807/196470999-b554507d-7883-40c5-8def-9546a5fc0c5d.png)

As for tests, I think just saying to open and close the floating action button a few times and verify that it happens quickly is a good test.

mountiny commented:
Thanks for the context Tim

hannojg committed:
Revert "perf: memo SidebarLinks"

This reverts commit 2d98ce0f6d44cd371f8bacfb49868f9ba5603b61.

hannojg committed:
refactor: moved PopoverMenu into its own component …

… instead of using memo. From performance perspective that even seems to have slightly improved over the React.memo approach.

hannojg committed:
clean: prop types

mountiny reviewed (commented):
Reassign @rushatgabhane for C+ review and testing as it seems like there was some Github blip which unassigned people from the PR.

> Do you recommend first creating issues, and then opening PRs (like, do you want us to have for every PR an issue)? 

Yes, let's always create an issue before you address it in a PR. You can use the standard or Performance issue template but please try to follow the template for the issues so it is easier for us to have the issues standardized. It is good to write down what metric is slow and we aim to improve with notes on why it is slow and plan in which we can improve this.

In the issue other people can also discuss what could be done. Some engineer can then assign you the issue so you can track it a bit easier.

mountiny commented:
Also regarding the performance comparison, you said you had to use this PR as base https://github.com/Expensify/App/pull/11886/files, however, can you then quantify what is the performance gain of this particular change (with no effect of the base change)

hannojg commented:
Alright, thank you for explaining, will create issues now first from now on 💪!

Yeah, the performance results of the main branch VS base branch VS this branch:

<img width="1280" alt="Screenshot 2022-10-20 at 2 50 47 PM" src="https://user-images.githubusercontent.com/16821682/196953674-1b857003-7752-4a55-8f9f-00f3a9aba15f.png">

will create a performance issue for both PRs soon!

mountiny reviewed (commented):
Thanks for looking into creating the issues. Code looks good to me, @rushatgabhane would you be bale to give this a test on all platforms? Feel free to re-assign to other C+ if you have full plate 🙇

mountiny reviewed (commented):
@hannojg will you be able to post in slack #expensify-margelo the issues that need to be created, I think contributors cannot create issues on their own actually just to make sure the repo is not clogged up. Someone will create the issue for you.

Then can you please add test and QA steps into the PR template. Can be something simple just to make sure the flow this change touches work as expected, the performance gains are already measured so maybe we could include some steps how reviewer can run the same tests locally and compare if they got same/similar improvements.

hannojg commented:
@rushatgabhane yes you are right, it should definitely be done on release builds.

I just realised that the app's UI thread I release mode seems very busy and is oscillating between 20-35 FPS, thus the performance benefits in the release build aren't really measurable as that greater performance issue shadows them. (I have it on my todo to look into that)

hannojg commented on src/components/ScreenWrapper/BaseScreenWrapper.js:
```diff
@@ -106,10 +118,5 @@ BaseScreenWrapper.defaultProps = defaultProps;
 export default compose(
     withNavigation,
     withWindowDimensions,
-    withOnyx({
```
`withOnyx` will add the onyx value as a prop to the component _and defer the rendering of the component until the value is available_ to be passed onto the component.

So two issues:

- We rerender the component by passing it as a prop, although the information we get from this prop is not necessary for the rendering output. 
- We unnecessarily defer/block the rendering slightly although not needed

hannojg replied:
And this is the `BaseScreenWrapper`. So if that one re-renders all child will rerender, and that's a massive operation (usually blocking my JS thread for a few seconds)

hannojg replied:
However, I just realized we could make it less imperative by using `shouldComponentUpdate` and just ignore the modal is open prop for 🤔 Will investigate that

hannojg replied:
Refactored

mountiny commented:
@hannojg @rushatgabhane What is a state on this PR? Lets get it merged.

hannojg commented:
I am looking into refactoring it to make it cleaner. Will mark it as draft, and un-draft as soon as we are good again 😊

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into perf/memoize-sidebar

hannojg committed:
removed onyx subscribe, use shouldComponentUpdate

hannojg committed:
code style

hannojg commented:
> Yes, this would be really helpful. The tool name used for iOS, Android and Web would suffice

I am working already on some performance tools that we can use during development! These will also be available soon to everyone.

hannojg commented:
(lint test failing due to some weird error, can someone please retry it? 😅 )

mountiny commented on src/pages/home/sidebar/SidebarScreen/PopoverModal.js:
```diff
@@ -0,0 +1,170 @@
+
+    /**
+     * Method called either when:
+     * Pressing the floating action button to open the CreateMenu modal
```
```suggestion
     * - Pressing the floating action button to open the CreateMenu modal
```

mountiny commented on src/pages/home/sidebar/SidebarScreen/PopoverModal.js:
```diff
@@ -0,0 +1,170 @@
+    /**
+     * Method called either when:
+     * Pressing the floating action button to open the CreateMenu modal
+     * Selecting an item on CreateMenu or closing it by clicking outside of the modal component
```
```suggestion
     * - Selecting an item on CreateMenu or closing it by clicking outside of the modal component
```

mountiny reviewed (commented):
Nice just one nitpick about a format of a comment, @rushatgabhane would you be able to give this a thorough test? 

Also excited for this perf improvement.

mountiny commented:
Retried it, can you also updated the PR reviewer checklist?

hannojg committed:
Apply suggestions from code review

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

hannojg commented:
Sorry, what do you mean by that exactly? 😅 I have ticked all boxes 🤔

mountiny commented:
It was probably old checklist or something, I copied in the new one and it passed now.

tgolen reviewed (commented):
Cool, I really like how this refactoring cleaned things up.

AndrewGable reviewed (dismissed):
https://github.com/bamlab/android-performance-profiler/ is really cool, code looks great!

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into perf/memoize-sidebar

hannojg committed:
fix issue after merge

hannojg commented:
Resolved!

rushatgabhane reviewed (dismissed):
@mountiny LGTM! ✨

#### Web

https://user-images.githubusercontent.com/29673073/198926117-690ce2e6-10c3-4cb9-8002-1d0b9a026b9c.mov


#### mWeb

https://user-images.githubusercontent.com/29673073/198926179-a1274136-a2b8-4a2f-be7b-78f0dfce0b74.mov


https://user-images.githubusercontent.com/29673073/198926193-c28006ea-fed2-454b-88f0-590f36f9b155.mov



#### iOS

https://user-images.githubusercontent.com/29673073/198926258-d90469a7-d073-4b0b-b5a0-18934af59897.mov


#### Android

https://user-images.githubusercontent.com/29673073/198926295-6d3ed8c3-6898-4e69-82c5-56a29eda93e9.mp4


#### Desktop

https://user-images.githubusercontent.com/29673073/198926077-b16525e7-865b-44d3-8c7d-0201fd5d9019.mov


<details>
<summary><h4>PR Reviewer Checklist</h4>

</summary>

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

hannojg commented on ios/NewExpensify.xcodeproj/project.pbxproj:
```diff
@@ -750,7 +750,7 @@
 				ENABLE_STRICT_OBJC_MSGSEND = YES;
 				ENABLE_TESTABILITY = YES;
-				"EXCLUDED_ARCHS[sdk=iphonesimulator*]" = "";
+				"EXCLUDED_ARCHS[sdk=iphonesimulator*]" = i386;
```
oh, sorry for committing this, will remove it, was by accident 😅

hannojg replied:
reverted!

hannojg committed:
revert change

hannojg committed:
revert change

mountiny reviewed (approved):
Thank you very much @rushatgabhane and @hannojg, lets get this one out! Code looks good to me, excited for these performance improvements coming in

mountiny commented:
Gonna merge since Andrew and Tim both expressed they like the code above.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3363442837) to staging by @mountiny in version: 1.2.22-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

Julesssss commented:
Hey @hannojg. I'm not 100% sure about this, but it's possible this refactor introduced [this regression](https://github.com/Expensify/App/issues/12312) where the FAB button is not being shown on Android. Would you mind taking a look?

hannojg commented:
Sure, will have a look tomorrow! (Today is national holiday in AT)

Julesssss commented:
FYI as this regression blocks users from almost all features and blocks further deploys, I'm going to revert it. 

@hannojg would you mind resubmitting a PR once this issue s resolved? Thanks!

![Screenshot_20221101-115046](https://user-images.githubusercontent.com/10736861/199227068-e36c95d1-41fa-4782-a481-1f127c63c99c.png)

hannojg commented:
Yes, please revert, will put out a new PR. Will make sure such a major regression won't happen again (also a bit confused as I thought I'd tested it also extensively on Android).

Will be first thing tmr morning

Julesssss commented:
FYI, here's the follow-up [PR](https://github.com/Expensify/App/pull/12369).

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3379106779) to production by @Julesssss in version: 1.2.22-3 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
